base/shell/explorer/traydeskbtn.cpp

```diff
@@ -0,0 +1,382 @@
+            {
+                //TODO: DPI scaling - return logical-to-physical conversion of 24, not fixed value
+                return 24;
+            }
+            else
+                return 15;
+        }
+        else
+        {
+            INT CurMargin = IsHorizontal
+                ? (m_ContentMargins.cxLeftWidth + m_ContentMargins.cxRightWidth)
+                : (m_ContentMargins.cyTopHeight + m_ContentMargins.cyBottomHeight);
+            return max(16 + CurMargin, 18) + 6;
+        }
+    }
+    else
+    {
+        return max(2 * GetSystemMetrics(SM_CXBORDER) + GetSystemMetrics(SM_CXSMICON),
+                   2 * GetSystemMetrics(SM_CYBORDER) + GetSystemMetrics(SM_CYSMICON));
+    }
+}
+
+HRESULT CTrayShowDesktopButton::DoCreate(HWND hwndParent)
+{
+    const DWORD style = WS_CHILD | WS_VISIBLE | WS_CLIPSIBLINGS | BS_DEFPUSHBUTTON;
+    Create(hwndParent, NULL, NULL, style);
+
+    if (!m_hWnd)
+        return E_FAIL;
+
+    // Get desktop icon
+    bool bIconRetrievalFailed = ExtractIconExW(L"imageres.dll", -IDI_IMAGERES_DESKTOP, NULL, &m_icon, 1) == UINT_MAX;
+    if (bIconRetrievalFailed || !m_icon)
+        ExtractIconExW(L"shell32.dll", -IDI_SHELL32_DESKTOP, NULL, &m_icon, 1);
+
+    // Get appropriate size at which to display desktop icon
+    m_szIcon.cx = GetSystemMetrics(SM_CXSMICON);
+    m_szIcon.cy = GetSystemMetrics(SM_CYSMICON);
+
+    // Prep visual style
+    EnsureWindowTheme(TRUE);
+
+    // Get HWND of Taskbar
+    m_hWndTaskbar = ::GetParent(hwndParent);
+    if (!::IsWindow(m_hWndTaskbar))
+        return E_FAIL;
+
+    return S_OK;
+}
+
+LRESULT CTrayShowDesktopButton::OnClick(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    // The actual action can be delayed as an expected behaviour.
+    // But a too late action is an unexpected behaviour.
+    LONG nTime0 = m_nClickedTime;
+    LONG nTime1 = ::GetMessageTime();
+    if (nTime1 - nTime0 >= 600) // Ignore after 0.6 sec
+        return 0;
+
+    // Show/Hide Desktop
+    ::SendMessage(m_hWndTaskbar, WM_COMMAND, TRAYCMD_TOGGLE_DESKTOP, 0);
+
+    return 0;
+}
+
+// This function is called from OnLButtonDown and parent.
+VOID CTrayShowDesktopButton::Click()
+{
+    // The actual action can be delayed as an expected behaviour.
+    m_nClickedTime = ::GetMessageTime();
+    PostMessage(TSDB_CLICK, 0, 0);
+}
+
+LRESULT CTrayShowDesktopButton::OnLButtonUp(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    m_bPressed = FALSE;
+    ReleaseCapture();
+    Invalidate(TRUE);
+
+    POINT pt;
+    ::GetCursorPos(&pt);
+    if (PtInButton(&pt))
+        Click(); // Left-click
+    return 0;
+}
+
+LRESULT CTrayShowDesktopButton::OnLButtonDown(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    m_bPressed = TRUE;
+    SetCapture();
+    Invalidate(TRUE);
+    return 0;
+}
+
+LRESULT CTrayShowDesktopButton::OnSettingChanged(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
```

Comment:
This handler is for the [`WM_SETTINGCHANGE`](https://learn.microsoft.com/en-us/windows/win32/winmsg/wm-settingchange) message.
This message is sent whenever any system-wide setting may have changed.
Unconditionally calling `OnThemeChanged` seems a little bit overkill at first sight, so could you tell us why this was done so?
Isn"t it possible to do some filtering based on the value of the `wParam` value? Because, note that when `WM_SETTINGCHANGE` is received (and thus, this handler called), the meaning of the `wParam` and `lParam` parameters are completely different than those of [`WM_THEMECHANGED`](https://learn.microsoft.com/sr-cyrl-rs/windows/win32/winmsg/wm-themechanged)... (Granted, in that case these parameters are not defined/reserved, so the `OnThemeChanged()` should not use them as well, but we don't really know about that...)

Reply:
idk ask Katayama

Reply:
Passing the parameters unchanged to OnThemeChanged is definitely wrong. Pass 0 instead.

The way you normally handle WM_SETTINGCHANGE is, check wParam and lParam. If they match whatever you are interested in, then process. If they are both 0, it means you don't know what changed so you have to process as if it matched.

So, what do you really care about here? SystemMetrics? Font size?

Reply:
@HBelusca @whindsaks I...don't know? As far as I can recall, I was just following the example of what was already there. Thank you though, to both of you, for the clarification as to *why* this is incorrect.

I agree it should be addressed...but I lack the information necessary to do so - while metrics, colours, and font sizes all seem likely...if there are any other reasons, I wouldn't know of them. Given my limited understanding and relative lack of experience, going off of "well I compiled and ran it, and nothing seemed wrong at a glance" seems like begging for trouble. What say you two?

Additionally, with all that in mind... @katahiromz do you have any input regarding this question?

Reply:
Here `WM_SETTINGCHANGE` is intended to catch metrics changes. 
`WM_THEMECHANGED` will be sent in theme change that might change GUI colors and appearance.

Reply:
This code freedom can be appreciated.
I agree.

Reply:
I'm asking about correctness (since the theme-changed handling would now be called _**everytime**_  a gobal system-wide win32k setting changes...), not about freedom

Reply:
I'm not a pro but a hobby programmer without money.

Reply:
https://git.reactos.org/?p=reactos.git&a=search&h=HEAD&st=grep&s=WM_SETTINGCHANGE

Reply:
@HBelusca Here's an idea:

### Proposed change

```cpp
LRESULT CTrayShowDesktopButton::OnSettingChanged(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
{
    HandleVisualPreferenceChanged(); //instead of calling `OnThemeChanged` directly
    EnsureWindowTheme(TRUE);
    return 0; //instead of a return value meant for `WM_THEMECHANGED`
}

LRESULT CTrayShowDesktopButton::OnThemeChanged(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
{
    HandleVisualPreferenceChanged();
    return 0;
}

//None of the message handlers' parameters are passed along to this function, since we shouldn't be conflating `WM_SETTINGCHANGE`'s `wParam`/`lParam`/etc. with that of `WM_THEMECHANGED`
VOID CTrayShowDesktopButton::HandleVisualPreferenceChanged()
{
    HIGHCONTRAST hcInfo;
    hcInfo.cbSize = sizeof(hcInfo);
    if (SystemParametersInfo(SPI_GETHIGHCONTRAST, sizeof(hcInfo), &hcInfo, FALSE))
        m_highContrastMode = (hcInfo.dwFlags & HCF_HIGHCONTRASTON);

    if (m_hTheme)
    {
        ::CloseThemeData(m_hTheme);
        m_hTheme = NULL;
    }
    if (m_hFallbackTheme)
    {
        ::CloseThemeData(m_hFallbackTheme);
        m_hFallbackTheme = NULL;
    }

    EnsureWindowTheme(FALSE);

    Invalidate(TRUE);
}
```
Previously, `OnSettingChanged` was doing pointless extra work in a theoretically-unsafe way (e.g. if someone later updates `OnThemeChanged` to actually use `wParam`/`lParam`/etc.)

### Short version

With this change, there would still be pointless work being done, but it would at least remove the possible danger.

### Long version

I will concede that with this, we'd still be handling change to *every* global "win32k setting" which is probably unnecessary, but at least this way we'd no longer be blindly passing `wParam` and `lParam` and such to a handler meant for a completely different window message.

### Verdict?

Given the apparent lack of clarity on *why* we're handling `WM_SETTINGCHANGE` regardless of *what* setting has changed, would something like this be an acceptable resolution?

Reply:
Question:
- Why would `OnSettingChange()` call `EnsureWindowTheme(TRUE);` (TRUE: set theme), while that `OnSettingChange()` handler is called when settings change;
- but why would `OnThemeChanged()` call (implicitly via the Handle... function) `EnsureWindowTheme(FALSE);` (FALSE: don'tset theme) ?

Wouldn't this be the other way around? or?

Also, what would happen if `EnsureWindowTheme(TRUE);` is _**not**_ called in `OnSettingChange()` (but keeping the rest of the code as you suggested)? Would this introduce visual problems when you change themes around and/or elements sizes (like increasing/decreasing font sizes, border sizes, etc.)?
![image](https://github.com/reactos/reactos/assets/1969829/11ba9072-b79e-4bff-bf00-00a751ae42ab)

Reply:
Question:
- Why would `OnSettingChange()` call `EnsureWindowTheme(TRUE);` (TRUE: set theme), while that `OnSettingChange()` handler is called when settings change;
- but why would `OnThemeChanged()` call (implicitly via the Handle... function) `EnsureWindowTheme(FALSE);` (FALSE: don'tset theme) ?

Wouldn't this be the other way around? or?

Also, what would happen if `EnsureWindowTheme(TRUE);` is _**not**_ called in `OnSettingChange()` (but keeping the rest of the code as you suggested)? Would this introduce visual problems when you change themes around and/or elements sizes (like increasing/decreasing font sizes, border sizes, etc.)?
![image](https://github.com/reactos/reactos/assets/1969829/11ba9072-b79e-4bff-bf00-00a751ae42ab)

Reply:
To give some additional feedback on what @whindsaks and @katahiromz mentioned, I went through the list of `uiActions` values (what would be passed to the `wParam` of the [`WM_SETTINGCHANGE`](https://learn.microsoft.com/en-us/windows/win32/winmsg/wm-settingchange#parameters) message) listed in the [MSDN page of SystemParametersInfoW()](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-systemparametersinfow#parameters), and here's a list of the settings that might need to be handled (the others to be dropped):
```
The following are the accessibility parameters.
===============================================

SPI_SETFOCUSBORDERHEIGHT
SPI_SETFOCUSBORDERWIDTH
SPI_SETHIGHCONTRAST
SPI_SETLOGICALDPIOVERRIDE

No idea about SPI_SETDESKWALLPAPER ...

The following are the icon parameters.
======================================

All the SPI_SET* in the list?


The following are the window parameters.
========================================

SPI_SETBORDER
SPI_SETMINIMIZEDMETRICS
SPI_SETNONCLIENTMETRICS

I don't think that
SPI_SETDRAGHEIGHT , SPI_SETDRAGWIDTH
have to be handled.
```

----

Regarding colors, [`WM_SYSCOLORCHANGE`](https://learn.microsoft.com/en-us/windows/win32/gdi/wm-syscolorchange) may be useful as well.